File: platforms/android/runtime/core/src/main/java/io/nativeblocks/runtime/api/provider/block/NativeBlockProviderRegistry.kt
package io.nativeblocks.runtime.api.provider.block

internal object NativeBlockProviderRegistry {

    private val providerMap = mutableMapOf<String, NativeBlockProvider>()
    private val lock = Any()

    fun getOrCreate(instanceName: String): NativeBlockProvider {
        synchronized(lock) {
            return providerMap.getOrPut(instanceName) {
                NativeBlockProvider()
            }
        }
    }

    fun remove(instanceName: String) {
        synchronized(lock) {
            providerMap.remove(instanceName)
        }
    }

}

File: platforms/ios/runtime/Sources/NativeblocksRuntime/api/provider/block/NativeBlockProviderRegistry.swift
import Foundation

internal class NativeBlockProviderRegistry {
    private static var providerMap: [String: NativeBlockProvider] = [:]
    private static let lock = NSLock()

    static func getOrCreate(_ instanceName: String) -> NativeBlockProvider {
        lock.lock()
        defer { lock.unlock() }

        if let existingProvider = providerMap[instanceName] {
            return existingProvider
        } else {
            let newProvider = NativeBlockProvider()
            providerMap[instanceName] = newProvider
            return newProvider
        }
    }

    static func remove(_ instanceName: String) {
        lock.lock()
        defer { lock.unlock() }
        providerMap.removeValue(forKey: instanceName)
    }

}
